[YSQL] Foreign Data Wrapper support

radekg opened the pull request:
This PR adds support for `postgres_fdw`. This PR targets the following github issue: https://github.com/yugabyte/yugabyte-db/issues/5714. At the current stage, it is possible to do the following:

```
bash-4.2$ ysqlsh
ysqlsh (11.2-YB-2.7.3.0-b0)
Type "help" for help.

yugabyte=# create role tenant1 with login password 'tenant1' connection limit 10;
CREATE ROLE
yugabyte=# create database tenant1db owner = tenant1;

CREATE DATABASE
yugabyte=# create role tenant2 with login password 'tenant2' connection limit 10;
CREATE ROLE
yugabyte=# create database tenant2db owner = tenant2;
CREATE DATABASE
yugabyte=# \connect tenant2db
You are now connected to database "tenant2db" as user "yugabyte".
tenant2db=# set role tenant2;
SET
tenant2db=> create table table1tenant2 (rowid int, rowval text);
CREATE TABLE
tenant2db=> set role yugabyte;
SET
tenant2db=# grant select, insert on table table1tenant2 to yugabyte;
GRANT
tenant2db=# \connect tenant1db
You are now connected to database "tenant1db" as user "yugabyte".
tenant1db=# create extension postgres_fdw;
CREATE EXTENSION
tenant1db=# CREATE SERVER foreigndb_fdw FOREIGN DATA WRAPPER postgres_fdw OPTIONS (host '127.0.0.1', port '5433', dbname 'tenant2db');
CREATE SERVER
tenant1db=# GRANT USAGE ON FOREIGN SERVER foreigndb_fdw TO yugabyte;
GRANT
tenant1db=# CREATE USER MAPPING FOR yugabyte SERVER foreigndb_fdw OPTIONS (user 'tenant2', password 'tenant2');
CREATE USER MAPPING
tenant1db=# CREATE FOREIGN TABLE table1tenant2 (
tenant1db(#   rowid integer OPTIONS (column_name 'rowid'),
tenant1db(#   rowval text OPTIONS (column_name 'rowval')
tenant1db(# ) SERVER foreigndb_fdw
tenant1db-# OPTIONS (schema_name 'public', table_name 'table1tenant2');
CREATE FOREIGN TABLE
tenant1db=# select * from table1tenant2;
 rowid | rowval
-------+--------
(0 rows)

tenant1db=# insert into table1tenant2 (rowid, rowval) values (1, 'aaa');
INSERT 0 1
tenant1db=# select * from table1tenant2;
 rowid | rowval
-------+--------
     1 | aaa
(1 row)

tenant1db=# \connect tenant2db
You are now connected to database "tenant2db" as user "yugabyte".
tenant2db=# set role tenant2;
SET
tenant2db=> select * from table1tenant2;
 rowid | rowval
-------+--------
     1 | aaa
(1 row)

tenant2db=>
```

There are some immediate issues:

1. I am not yet able to `import foreign schema`:

```
tenant1db=# IMPORT FOREIGN SCHEMA public LIMIT TO (table1tenant2) FROM SERVER foreigndb_fdw INTO public;
ERROR:  COLLATE not supported yet
LINE 3:   rowval text OPTIONS (column_name 'rowval') COLLATE pg_cata...
                                                     ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1127. Click '+' on the description to raise its priority
QUERY:  CREATE FOREIGN TABLE table1tenant2 (
  rowid integer OPTIONS (column_name 'rowid'),
  rowval text OPTIONS (column_name 'rowval') COLLATE pg_catalog."default"
) SERVER foreigndb_fdw
OPTIONS (schema_name 'public', table_name 'table1tenant2');
```

~~It seems that I was not able to locate the correct unsupported `COLLATE` case (https://github.com/YugaByte/yugabyte-db/issues/1127).~~ This one is due to this: https://github.com/yugabyte/yugabyte-db/blob/v2.7.2/src/postgres/src/backend/parser/gram.y#L3722, as an interim solution I can configure and compile YugabyteDB with `YB_POSTGRES_WITH_ICU=[true | 1 | t | yes | y]` environment variable, **then it all works**:

```
tenant1db=# IMPORT FOREIGN SCHEMA public LIMIT TO (table1tenant2) FROM SERVER foreigndb_fdw INTO public;
IMPORT FOREIGN SCHEMA
```

2. Not sure if this is a real issue: 

```
tenant1db=> IMPORT FOREIGN SCHEMA public LIMIT TO (table1tenant2) FROM SERVER foreigndb_fdw INTO public;
ERROR:  password is required
DETAIL:  Non-superuser cannot connect if the server does not request a password.
HINT:  Target server's authentication method must be changed.
```

3. Currently, the `postgres_fdw` extension has been hard coded in the `shared_preload_libraries` configuration directive. Preferably, this should be supported as a user configurable switch. However, for this to be the case, the following PR would have to go in first: https://github.com/yugabyte/yugabyte-db/pull/9576.
4. Regression tests have to be adapted.
5. I would assume some a test suite more comprehensive than a regression test suite might be needed.
6. Obviously, there might be some limitations on the DocDB side leading to this not working in a distributed environment, which I am not aware of.
7. Needs to be tested in a distributed cluster. Currently tested using the single node launched via `yugabyted`.

Early feedback welcomed.

radekg commented on src/postgres/src/backend/parser/gram.y:
```diff
@@ -975,9 +975,9 @@ stmt :
-			| AlterFdwStmt { parser_ybc_not_support(@1, "This statement"); }
-			| AlterForeignServerStmt { parser_ybc_not_support(@1, "This statement"); }
-			| AlterForeignTableStmt { parser_ybc_not_support(@1, "This statement"); }
+			| AlterFdwStmt
```
The new features from this list should go to the list of beta features.

fizaaluthra replied:
Yeah I believe the new features should be moved to that list now

radekg replied:
Moved all the features to the beta section: https://github.com/yugabyte/yugabyte-db/pull/9650/commits/940f02190359656406e4dd58f2b7c72d9fa9716d.

radekg commented on src/postgres/src/backend/parser/gram.y:
```diff
@@ -8200,9 +8174,7 @@ index_including_params:	index_elem						{ $$ = list_make1($1); }
 			| index_including_params ',' index_elem		{ $$ = lappend($1, $3); }
 		;
 
-opt_collate: COLLATE any_name						{
```
This is wrong and has to be restored.

radekg commented on src/postgres/src/pl/plpgsql/src/pl_gram.y:
```diff
@@ -798,19 +798,16 @@ decl_collate	:
 					{ $$ = InvalidOid; }
 				| K_COLLATE T_WORD
 					{
-						ybc_not_support(@1, "COLLATE", 1127);
```
These do not appear to be the correct `COLLATE` cases required by the `IMPORT FOREIGN SCHEMA`.

radekg commented on src/yb/yql/pgwrapper/pg_wrapper.cc:
```diff
@@ -353,9 +353,9 @@ Status PgWrapper::Start() {
   if (FLAGS_pg_stat_statements_enabled) {
     argv.push_back("shared_preload_libraries=pg_stat_statements,yb_pg_metrics,pgaudit,"
-      "pg_hint_plan");
+      "pg_hint_plan,postgres_fdw");
```
This should use the method from: https://github.com/yugabyte/yugabyte-db/pull/9576.

radekg commented on src/yb/yql/pgwrapper/pg_wrapper.cc:
```diff
@@ -353,9 +353,9 @@ Status PgWrapper::Start() {
+      "pg_hint_plan,postgres_fdw");
   } else {
-    argv.push_back("shared_preload_libraries=yb_pg_metrics,pgaudit,pg_hint_plan");
+    argv.push_back("shared_preload_libraries=yb_pg_metrics,pgaudit,pg_hint_plan,postgres_fdw");
```
Same as above.

radekg commented on src/yb/yql/pgwrapper/pg_wrapper.cc:
```diff
@@ -371,6 +371,11 @@ Status PgWrapper::Start() {
+    GetPostgresLibPath(),
+    GetPostgresThirdPartyLibPath()
+  };
+  pg_proc_->SetEnv("LD_LIBRARY_PATH", boost::join(ld_library_path, ":"));
```
This is required for the `postgres_fdw` extension to load. Without this, the following happens:

```
yugabyte=# create extension postgres_fdw;
ERROR:  could not load library "/home/yb/postgres/lib/postgres_fdw.so": libpq.so.5: cannot open shared object file: No such file or directory
```

radekg commented on src/postgres/src/test/regress/expected/yb_pg_foreign_data.out:
```diff
@@ -0,0 +1,2268 @@
+RESET ROLE;
+DROP ROLE regress_test_indirect;                            -- ERROR
+ERROR:  role "regress_test_indirect" cannot be dropped because some objects depend on it
+DETAIL:  owner of server s1
```
I have no idea why but this has a tendency to arrive in a different order.

radekg commented on src/postgres/src/test/regress/expected/yb_pg_foreign_data.out:
```diff
@@ -0,0 +1,2268 @@
+DROP SCHEMA foreign_schema CASCADE;
+DROP ROLE regress_test_role;                                -- ERROR
+ERROR:  role "regress_test_role" cannot be dropped because some objects depend on it
+DETAIL:  privileges for foreign-data wrapper foo
```
I have no idea why but this has a tendency to arrive in a different order.

This + the `DETAIL` at line 444 and 445 makes this test fail almost every time.

radekg replied:
I think the reason for this is hidden in the implementation of `ybc_systable_beginscan`. The reason why I think this is the case is:

- the error message `cannot be dropped because some objects depend on it` exists in [`user.c`](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/backend/commands/user.c#L1067)
- which calls [`checkSharedDependencies`](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/backend/catalog/pg_shdepend.c#L553)
- that function works on the result of [`systable_beginscan`](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/backend/access/index/genam.c#L334) which in case of YugabyteDB is redefined to:

```c
	if (IsYugaByteEnabled())
	{
		return ybc_systable_beginscan(heapRelation,
		                              indexId,
		                              indexOK,
		                              snapshot,
		                              nkeys,
		                              key);
	}
```

My assumption is: Postgres works on an ordered _scan_ while YugabyteDB implementation appears to be working on a random order _scan_.

radekg replied:
To me, the most reasonable approach here would be to string the detail string by `\n` and sort the resulting array, then join back. It may not be fully compatible with Postgres but at least the order will be predictive. Unless this problem surfaces some underlying index storage problem...

radekg commented:
There are following problems:

- https://github.com/yugabyte/yugabyte-db/pull/9650#discussion_r687249211: `DETAIL` arrives at a random order, even when I run this one java test with `--scb --sj`
- https://github.com/yugabyte/yugabyte-db/pull/9650#discussion_r687249504: same as above

These detail issues are the only thing stopping this test from succeeding... Other than that, everything seems to be dandy. Does anyone have an idea why the `DETAIL`  is arriving in random order? It seems that my test is the only test receiving the `DETAIL`  with error messages so I have no reference to base myself on.

However, looking at what's passing in here:

- servers can be created / altered / dropped
- import foreign schema works, I know there's an issue when YB is configured without ICU because postgres_fdw sticks in a `COLLATE` in the create foreign table
- foreign tables can be created and dropped, altering has the same constraints as regular alter table
- insert / selects via foreign tables also works and obeys the postgres rbac

I'd love some feedback.

I think my next steps are:
- adapt existing tests
- it would be really awesome is someone could have a look at this related PR: https://github.com/yugabyte/yugabyte-db/pull/9576

Thank you!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/yugabyte/yugabyte-db?pullRequest=9650) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/yugabyte/yugabyte-db?pullRequest=9650) before we can accept your contribution.<br/>**16** out of **25** committers have signed the CLA.<br/><br/>:white_check_mark: ddhodge<br/>:white_check_mark: jaki<br/>:white_check_mark: radekg<br/>:white_check_mark: ttyusupov<br/>:white_check_mark: lizayugabyte<br/>:white_check_mark: bllewell<br/>:white_check_mark: tedyu<br/>:white_check_mark: vmallepalli<br/>:white_check_mark: yb-andrew<br/>:white_check_mark: mbautin<br/>:white_check_mark: tanujnay112<br/>:white_check_mark: atarw<br/>:white_check_mark: deeps1991<br/>:white_check_mark: d-uspenskiy<br/>:white_check_mark: iSignal<br/>:white_check_mark: polarweasel<br/>:x: spolitov<br/>:x: hulien22<br/>:x: mchiddy<br/>:x: daniel-yb<br/>:x: sb-yb<br/>:x: anmalysh-yb<br/>:x: sunilsutarDCE<br/>:x: hkandala<br/>:x: SergeyPotachev<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/yugabyte/yugabyte-db?pullRequest=9650) it.</sub>

radekg commented:
Updated existing tests with FDW related features:

- src/postgres/src/test/regress/sql/yb_pg_event_trigger.sql
- src/postgres/src/test/regress/sql/yb_pg_rolenames.sql
- src/postgres/src/test/regress/sql/yb_pg_alter_generic.sql

These are passing.

radekg commented on java/yb-pgsql/src/test/java/org/yb/pgsql/TestPgRegressForeignData.java:
```diff
@@ -0,0 +1,33 @@
+  public void testPgRegressForeignData() throws Exception {
+    runPgRegressTest("yb_pg_foreign_data_schedule");
+  }
+}
```
Done: https://github.com/yugabyte/yugabyte-db/pull/9650/commits/4935c5ae955b40f00d93eb14a27aac5d0b2fa36d.

radekg commented on src/postgres/src/test/regress/expected/yb_pg_foreign_data.out:
```diff
@@ -0,0 +1,2268 @@
+ umuser | umserver | umoptions 
+--------+----------+-----------
+(0 rows)
+
```
Do you have a preference how this should be split? I have followed the one to one mapping by moving the forward_data.out to yb_pg_forward_data.out and there's always one to one. Preferably, once the problem with DETAIL coming out in random order is solved, this test will preferably not fail anymore.

fizaaluthra replied:
Given that there's other ways to view the diff between expected o/p and actual o/p, I'm good with the one to one mapping, thanks!

radekg commented:
Now that https://github.com/yugabyte/yugabyte-db/pull/9576 has been merged, this PR can be tested by using the `--ysql_pg_conf_csv="\"shared_preload_libraries='postgres_fdw'\""`. Hey @fizaaluthra and @m-iancu, looking forward to further feedback on this one.

The most urgent one is the random `DETAIL` order, this is the reason why the _foreign\_data_ test is failing.

m-iancu commented:
>The most urgent one is the random DETAIL order, this is the reason why the foreign_data test is failing.

@radekg I have not looked at the issue regarding the DETAIL order carefully, but if my understanding is correct, then I think we can just suppress the detail lines by calling  `\set VERBOSITY terse` before the critical command and then `\set VERBOSITY default` after to reset. 

We do this in several other tests such as: 
* [yb_pg_triggers.out#L585](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/test/regress/expected/yb_pg_triggers.out#L585)
* [yb_pg_foreign_key.out#L1919](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/test/regress/expected/yb_pg_foreign_key.out#L1919)

See also [yb_pg_event_trigger.out#L166](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/src/test/regress/expected/yb_pg_event_trigger.out#L166) which has a bit of an explanation of the root of the issue.

>Do you have a preference how this should be split? 

We already have some pretty big test files, so if there's no obvious way to split them I think we can keep it as it is. 
But if you did it already that's also fine. cc @fizaaluthra 

Note: For debugging test failures, it's usually easier to just go to the regress build folder (e.g. `~/code/yugabyte-db/build/latest/postgres_build/src/test/regress`, and there use some diffing tool to compare results vs expected.
The test output will also output some comparison but that's usually harder to read.
For example:
```
$ cd ~/code/yugabyte-db/build/latest/postgres_build/src/test/regress
$ vimdiff expected/yb_pg_foreign_data.out results/yb_pg_foreign_data.out 
```

radekg commented:
Hey @m-iancu. Thank you for looking into this. Normally, I would be inclined to go for setting the verbosity but... I do have a feeling this is not a good idea. Let me quickly explain. Let's look at the details:

At line 444:

```
ERROR:  role "regress_test_role" cannot be dropped because some objects depend on it
DETAIL: owner of server s1
privileges for foreign-data wrapper foo
```

and line 2224:

```
ERROR:  role "regress_test_role" cannot be dropped because some objects depend on it
DETAIL:  privileges for foreign-data wrapper foo
owner of user mapping for regress_test_role on server s6
privileges for server s4
```

The DETAIL section explicitly mentions which objects disallow dropping the role. Skipping these messages by changing the  verbosity poetntially opens the database for a very nasty to track regression problem. If, for whatever reason, the dependency checking would go wrong, it becomes untraceable in the test.

The question I am asking myself is: if this detail is not important, why would the original Postgres authors go to the extent of testing these detail messages instead of switching the verbosity? I do have the feeling that the order of the DETAIL message is irrelevant but it should be consistent across executions. Sorting the lines alphabetically (and rejecting empty lines, which do happen!) is the cheap way to go here.

I haven't split the big test and I would prefer to keep it that way so it maps nicely to the original Postgres tests.

Regarding the test result validation: on a failed test, the Java runner creates a bunch of mojo xml file, one of them is (AFAIR) mojo ending with _-result.xml_. Opening that in vi gives a full log of the failed test with the nicely formatted _expected_ and _out_ side by side where the diffs are marked with `>`, `<` and `|` markers. That's what I use to validate the diffs.

Let me know what your thoughts are on the DETAIL issue.

fizaaluthra commented on src/postgres/src/test/regress/yb_pg_foreign_data_schedule:
```diff
@@ -0,0 +1,10 @@
+# Dependency for `yb_pg_foreign_data`
+test: yb_create_function
+
+test: yb_pg_foreign_data
```
nit: newline at the end of file

radekg replied:
Resolved: https://github.com/yugabyte/yugabyte-db/pull/9650/commits/ce0f8924c9c49dca9ea9279912a3ac815cb35917

fizaaluthra commented:
For the DETAIL Issue - i like the idea of sorting the lines. LGTM otherwise! Thank you @radekg :)

radekg commented:
I have following failures but none of them appear to be immediately related to the new code:

```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestIndexBackfill.insertsWhileCreatingIndex:150->BasePgSQLTest.assertQuery:1233->BasePgSQLTest.assertRows:1190->BasePgSQLTest.assertRows:1198->BasePgSQLTest.assertRow:1212 Mismatch at row 1: Column #1 mismatch: expected:<500> but was:<499>
Expected row: Row[0=java.lang.Integer::500]
Actual row:   Row[count=java.lang.Long::499]
[ERROR]   TestPgColumnReadEfficiency.testScans:67->checkIndexOnlyScan:120->checkQuery:99 Expected median time 2565.019000 of 'SELECT v1, v2, v3, t1, t2, t3, t4, t5 FROM t WHERE v1 = 1' to be greater than median time 2587.951000 of 'SELECT v2 FROM t WHERE v1 = 1'
[ERROR]   TestPgPrefetchControl.testLimitPerformance:223->BasePgSQLTest.assertQueryRuntimeWithRowCount:1579 Query 'org.postgresql.jdbc.PgStatement@24cba153' took 586 ms! Expected 500 ms at most
[ERROR]   TestPgPushdown.inequalityAndIn_allRangeColumns:538->BasePgSQLTest.assertQueryRuntimeWithRowCount:1579 Query 'org.postgresql.jdbc.PgStatement@580435c3' took 4729 ms! Expected 3000 ms at most
[ERROR]   TestPgPushdown.inequalityAndIn_twoRangeColumns:198->BasePgSQLTest.assertQueryRuntimeWithRowCount:1579 Query 'org.postgresql.jdbc.PgStatement@73bff02c' took 4991 ms! Expected 3000 ms at most
[ERROR]   TestPgPushdown.inequality_oneRangeColumn:102->BasePgSQLTest.assertQueryRuntimeWithRowCount:1579 Query 'org.postgresql.jdbc.PgStatement@665d650' took 5125 ms! Expected 2500 ms at most
[ERROR]   TestPgPushdown.inequality_twoRangeColumns:443->BasePgSQLTest.assertQueryRuntimeWithRowCount:1579 Query 'org.postgresql.jdbc.PgStatement@3836685b' took 4611 ms! Expected 2500 ms at most
[ERROR]   TestPgRegressAuthorization.testPgRegressAuthorization:31->BasePgSQLTest.runPgRegressTest:138->BasePgSQLTest.runPgRegressTest:134->BasePgSQLTest.runPgRegressTest:125 pg_regress exited with error code: 1, failed tests: [yb_pg_rowsecurity]
[ERROR]   TestPgRegressTypesUDT.testPgRegressTypesUDT:31->BasePgSQLTest.runPgRegressTest:138->BasePgSQLTest.runPgRegressTest:134->BasePgSQLTest.runPgRegressTest:125 pg_regress exited with error code: 1, failed tests: [yb_create_function]
[ERROR] Errors:
[ERROR]   TestPgRegressLargeTable.testPgRegressLargeTable:114->BasePgSQLTest.assertQueryRuntimeWithRowCount:1578->BasePgSQLTest.timeQueryWithRowCount:1492->BasePgSQLTest.runQueryWithRowCount:1446 ? TestTimedOut
[ERROR]   TestTransactionStatusTable.lambda$testCreation$0:52->BasePgSQLTest.assertOneRow:1260 ? PSQL
[ERROR]   TestYbBackup.testAlteredTable:110->doAlteredTableBackup:81 ? YBBackup Failed p...
[ERROR]   TestYbBackup.testAlteredTableInColocatedDB:116->doAlteredTableBackup:81 ? YBBackup
[ERROR]   TestYbBackup.testAlteredTableInOriginalCluster:224 ? YBBackup Failed process w...
[ERROR]   TestYbBackup.testAlteredTableWithNotNull:307 ? YBBackup Failed process with ex...
[ERROR]   TestYbBackup.testColocatedWithTableOidAlreadySet:265 ? YBBackup Failed process...
[ERROR]   TestYbBackup.testIndex:344 ? YBBackup Failed process with exit code 1: [/yb-so...
[ERROR]   TestYbBackup.testIndexTypes:394 ? YBBackup Failed process with exit code 1: [/...
[ERROR]   TestYbBackup.testMixedColocatedDatabase:149 ? YBBackup Failed process with exi...
[ERROR]   TestYbBackup.testPartialIndexes:463 ? YBBackup Failed process with exit code 1...
[ERROR]   TestYbBackup.testRestoreWithRestoreTime:498 ? YBBackup Failed process with exi...
[INFO]
[ERROR] Tests run: 515, Failures: 9, Errors: 12, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10:37 h
[INFO] Finished at: 2021-08-28T12:15:25Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project yb-pgsql: There are test failures.
[ERROR]
[ERROR] Please refer to /yb-source/java/yb-pgsql/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

fizaaluthra commented:
Hey @radekg, with the latest version of the code I think we're good to replace the `malloc`, `realloc`, `free` with `palloc`, `repalloc`, `pfree`. I tested it out and it seems to work fine with the FDW test. 
Ideally we want to comply with the postgres code. 
Thanks!

radekg commented:
Hey @fizaaluthra, will check this out tomorrow. Why would it work now? Is there a commit which fixes that? Would you be able to point to it? Just for learning purposes :)

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,130 @@ yb_hash_code(PG_FUNCTION_ARGS)
+static int default_detail_store_capacity = 10;
+
+typedef struct {
+	char ** lines;
```
nit: no space  after **

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,130 @@ yb_hash_code(PG_FUNCTION_ARGS)
+	v->capacity = default_detail_store_capacity;
+	v->lines = (char **)malloc(sizeof(char *) * v->capacity);
+	v->length = 0;
+}
```
nit: another newline after each function 
same for the functions below

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,130 @@ yb_hash_code(PG_FUNCTION_ARGS)
+	}
+	return 0;
+}
+char * detailStoreGetAt(DetailStore *v, int index)
```
nit: no whitespace after * 
same with line 1538 and line 1593

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,130 @@ yb_hash_code(PG_FUNCTION_ARGS)
+void initDetailStore(DetailStore *v)
+{
+	v->capacity = default_detail_store_capacity;
+	v->lines = (char **)malloc(sizeof(char *) * v->capacity);
```
After commit `683f313`, replacing malloc / realloc / free with palloc / repalloc and pfree works fine, so we can replace these.

fizaaluthra reviewed (commented):
looks good! 
There's something off about the git commit history -- when I try to rebase with master there are merge conflicts but the conflicting code is already duplicated into this branch

radekg commented:
Hi @fizaaluthra. I have fixed the branch by branching out again from master and cherry picking commits from the botched branch. Fixed all spacing issues and I did not like the fixed default capacity with occasional `repalloc` so I have reimplemented the whole thing to use the Postgres `List` type. This looks much better now and there's less code to look at.

radekg commented:
The CI failure does not seem to be related to my changes at all.

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,99 @@ yb_hash_code(PG_FUNCTION_ARGS)
+	int length;
+} DetailSorter;
+
+void detailSorterFromList(DetailSorter *v, List *litems, int capacity)
```
I think the naming convention is that these helper functions should be in underscore case (`detail_sorter_from_list`)

radekg replied:
Hah, I was confused by this. `List` functions use that type of casing while the `StringInfo` uses the other. Addressed in this commit: https://github.com/yugabyte/yugabyte-db/pull/9650/commits/96eeb6b8a7253ae54952896fb54cb41ea21db32d.

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1474,64 @@ yb_hash_code(PG_FUNCTION_ARGS)
+    return strcmp(*(const char **) a, *(const char **) b);
+}
+
+char *yb_detail_sorted(char *input)
```
nit: This should be in PascalCase like `YBDetailSorted`

fizaaluthra commented on src/postgres/src/backend/utils/misc/pg_yb_utils.c:
```diff
@@ -1474,3 +1476,99 @@ yb_hash_code(PG_FUNCTION_ARGS)
+	}
+	else
+	{
+		char ** sortedLines = detailSorterLinesSorted(&sorter);
```
nit: no space after `**` and underscore case for `sorted_lines`

radekg replied:
https://github.com/yugabyte/yugabyte-db/pull/9650/commits/96eeb6b8a7253ae54952896fb54cb41ea21db32d

fizaaluthra reviewed (commented):
Awesome!! I'll do a final run through our test suite after this pass and I think we should be good to go after :)

radekg commented:
All addressed.

fizaaluthra commented:
Cool! I put it through the test suite - will approve it after :)

fizaaluthra commented on src/postgres/src/test/regress/yb_pg_foreign_data_schedule:
```diff
@@ -0,0 +1,10 @@
+# This includes tests related to postgres foreign data.
+####################################################################################################
+# Dependency for `yb_pg_foreign_data`
+test: yb_create_function
```
Hey @radekg, currently `TestPgRegressTypesUDT` is failing because the test executes `yb_pg_create_function_1.source` and `yb_create_function` - both of which now create the test_fdw_handler function and so the second `CREATE FUNCTION` throws an error. 
We can add `yb_pg_create_function_1` to the schedule here instead

fizaaluthra commented on src/postgres/src/test/regress/input/yb_create_function.source:
```diff
@@ -11,3 +11,8 @@ CREATE FUNCTION bigname_out(bigname)
    AS '@libdir@/regress@DLSUFFIX@'
    LANGUAGE C STRICT IMMUTABLE;
+
+CREATE FUNCTION test_fdw_handler()
```
and we can remove this

fizaaluthra commented on src/postgres/src/test/regress/output/yb_create_function.source:
```diff
@@ -12,3 +12,7 @@ CREATE FUNCTION bigname_out(bigname)
    AS '@libdir@/regress@DLSUFFIX@'
    LANGUAGE C STRICT IMMUTABLE;
 NOTICE:  argument type bigname is only a shell
+CREATE FUNCTION test_fdw_handler()
```
and this

radekg commented:
Running the tests with updated schedules right now.

fizaaluthra reviewed (approved):
LGTM!!! :)

radekg commented:
Very cool, thank you!
